### DISCUSSION TITLE
Wiki contribution

### DISCUSSION THREAD

**Discussion by Mathadon** [john30/ebusd#781] [Show and tell]
I had some difficulty finding my way through the documentation and GitHub issues trying to set set points for my boiler. I would like to contribute my findings of how I eventually got things working and important steps in the process, such that other can apply it to their case/device. My suggestion would be to include this post as a wiki page and I'd be glad to further clarify in this discussion in case it is not clear.

## Goal
I want to control my Bulex (Saunier Duval) Thermomaster: be able to set its supply water set point for the heating system of my house. The heater supports `ebus` so I ended up on this Github repository a few years ago but only now managed to find the time to dig a bit deeper and finally got things working.

## Hardware 
I bought an [Esara ebusd USB coupler](https://esera.de/en/Products/eBus/135/eBus-Coupler-USB) and 'tuned' the potentiometer a few years ago. I can't comment on the process since I don't recall. In the meantime it seems however that this device is no longer needed since a raspberry Pi's GPIO pins can be used as well. See more about this on [this wiki page](https://github.com/john30/ebusd/wiki/eBus-with-Raspberry-Pi-Serial). I did not test this yet but I intend to in the (perhaps very distant) future.

Edit 9/3/2023, caution, only more or less accurate:
> I tested the GPIO pin and it seems that a raspberry PI uses the wrong voltage levels. By default raspberry pi uses 0 to 3.3V for its UART while eBus uses 9 V to 22 V. So they are incompatible unless you convert the voltage levels. Using my (limited) knowledge of electronics this could be easy to do for reading from the bus, using a voltage divider. However for writing on the bus things become more complex. I decided to have my own 'eBuzzz' adapter produced by JLCPBC using the schematics from https://gitlab.com/fromeijn/ebuzzz-adapter . It costs about 130 EUR for 5 incl. shipping. Some components were out of stock but it was quite straight forward to find substitutes with (nearly) the same electrical properties and sizing using the component selector.

I connected the USB coupler to my boiler, which roughly consisted of the following steps:
 - Remove the boiler front panel
 - Open up the circuit board panel
 - Connect 2 wires of a regular UTP cable to the same port as the port which the thermostat is connected to.
 - Connect the other end to the Esara coupler.

Next, I coupled the USB port of the coupler to a Raspberry pi. After logging in to the raspberry pi through ssh (or in a different way), the coupler should now be listed under `ls /dev/ttyUSB0`. 

## Installing ebusd
Install libtool if you didn't do so before: 

`sudo apt-get update && sudo apt-get install libtool`

The following steps were taken to install `ebusd`, based on the [documentation](https://github.com/john30/ebusd).

```bash
git clone https://github.com/john30/ebusd.git
cd ebusd
./autogen.sh
sudo make install-strip
```

## Starting ebusd
At this point ebusd is installed but not running. It took a while before I understood that `ebusd` is intended to run in the background: you start up a process using the command `ebusd` and you can communicate with that 'daemon' using tcp using the command `ebusctl`. Furthermore, the ebusd log can be inspected at `/var/log/ebusd.log` and contains a log of the information that passes on the bus.

Ebusd can be started from the command line and requires some parameters/arguments. Defaults for these parameters can be set using environment variables or in the file `/etc/default/ebusd` but somehow that file is ignored for me. I still have to pass all the arguments to `ebusd`. The following set of arguments worked for what I wanted to achieve:

```
sudo ebusd -d /dev/ttyUSB0 --enablehex --scanconfig=full
```

Some clarification: `-d /dev/ttyUSB0` points ebusd to the device I want to use. `--enablehex` enables a mode where you can directly write hexadecimal commands to the bus. I needed `--scanconfig=full` instead of `--scanconfig` to get my device detected. [Scanconfig documentation here](https://github.com/john30/ebusd/wiki/2.-Run#message-configuration-options).

> Note: Executing multiple ebusd calls in a row will stop the previously ebusd call from running. You will then see something like this in the log
> ```
> 2022-12-29 14:06:37.451 [main error] can't open pidfile: /var/run/ebusd.pid, exiting
> 2022-12-29 14:06:37.452 [main notice] ebusd stopped
> ```
> This is _not_ an indication that ebusd failed or that there are writing permission problems or anything like that. It's just ebusd telling you that it's shutting down correctly since you asked for it to shut down before starting a new instance.

## Scanning the bus

At this point you should be able to run:
```
ebusctl info
```

and get an output like this:

```
update check: revision v22.4 available
device: /dev/ttyUSB0
signal: acquired
symbol rate: 23
max symbol rate: 98
min arbitration micros: 410
max arbitration micros: 455
min symbol latency: 5
max symbol latency: 6
reconnects: 0
masters: 2
messages: 224
conditional: 3
poll: 0
update: 10
address 03: master #11
address 08: slave #11, scanned "MF=Vaillant;ID=BAI00;SW=0604;HW=1303", loaded "vaillant/bai.308523.inc", "vaillant/08.bai.csv"
address 10: master #2
address 15: slave #2, scanned "MF=Vaillant;ID=E7C00;SW=0206;HW=7402", loaded "vaillant/15.e7c.csv"
address 31: master #8, ebusd
address 36: slave #8, ebusd, scanning
```

However, if you didn't wait long enough after starting `ebusd`, it didn't have time yet to detect which devices were on the bus, and you can get something like this:
```
version: ebusd 22.4.v22.4-57-g5d3c6f09
device: /dev/ttyUSB0
signal: acquired
symbol rate: 23
max symbol rate: 24
reconnects: 0
masters: 1
messages: 11
conditional: 0
poll: 0
update: 4
address 31: master #8, ebusd
address 36: slave #8, ebusd
```
You get the same output when not using `--scanconfig=full`. The log does not show which type of devices are connected and that's not a good sign.

## Inspecting the log
At his point it's useful to have a look at the log, which may look like this:
```
2022-12-29 14:24:14.119 [main notice] ebusd 22.4.v22.4-57-g5d3c6f09 started with full scan on device /dev/ttyUSB0
2022-12-29 14:24:14.440 [bus notice] bus started with own address 31/36
2022-12-29 14:24:14.441 [bus notice] signal acquired
2022-12-29 14:24:15.806 [bus notice] new master 10, master count 2
2022-12-29 14:24:15.832 [bus notice] new master 03, master count 3
2022-12-29 14:24:15.832 [update notice] received unknown MS cmd: 1008b510090000b476ffff000100 / 0101
2022-12-29 14:24:20.839 [update notice] received unknown MS cmd: 1008b5110101 / 09825a0080ffff0100ff
2022-12-29 14:24:22.849 [update notice] received unknown MS cmd: 1008b5040100 / 0a00ffffffffffffff0080
2022-12-29 14:24:23.110 [update notice] received unknown MS cmd: 1008b5110102 / 0601148c466e50
2022-12-29 14:24:24.441 [main notice] starting initial full scan
2022-12-29 14:24:27.833 [bus notice] scan 08: ;Vaillant;BAI00;0604;1303
2022-12-29 14:24:27.833 [update notice] store 08 ident: done
2022-12-29 14:24:27.833 [update notice] sent scan-read scan.08  QQ=31: Vaillant;BAI00;0604;1303
2022-12-29 14:24:27.833 [bus notice] scan 08: ;Vaillant;BAI00;0604;1303
2022-12-29 14:24:31.322 [update notice] received unknown MS cmd: 1008b5110101 / 09835a0080ffff0100ff
2022-12-29 14:24:32.528 [bus notice] scan 15: ;Vaillant;E7C00;0206;7402
2022-12-29 14:24:32.529 [update notice] store 15 ident: done
2022-12-29 14:24:32.529 [update notice] sent scan-read scan.15  QQ=31: Vaillant;E7C00;0206;7402
2022-12-29 14:24:32.529 [bus notice] scan 15: ;Vaillant;E7C00;0206;7402
2022-12-29 14:24:33.187 [update notice] received unknown MS cmd: 1008b512020000 / 00
2022-12-29 14:24:34.758 [main notice] read common config file vaillant/broadcast.csv
2022-12-29 14:24:34.829 [main notice] read common config file vaillant/scan.csv
2022-12-29 14:24:34.900 [main notice] read common config file vaillant/general.csv
2022-12-29 14:24:34.978 [main notice] read scan config file vaillant/08.bai.csv for ID "bai00", SW0604, HW1303
2022-12-29 14:24:36.025 [update notice] received unknown MS cmd: 1008b510090000b476ffff000100 / 0101
2022-12-29 14:24:36.689 [update notice] sent scan-read scan.08 id QQ=31: 
2022-12-29 14:24:37.398 [update notice] sent scan-read scan.08 id QQ=31: 
2022-12-29 14:24:38.108 [update notice] sent scan-read scan.08 id QQ=31: 
2022-12-29 14:24:38.820 [update notice] sent scan-read scan.08 id QQ=31: 21;16;12;0010016667;1610;005167;N6
2022-12-29 14:24:39.125 [main notice] found messages: 225 (3 conditional on 28 conditions, 0 poll, 10 update)
2022-12-29 14:24:40.078 [update notice] sent scan-read scan.08 id QQ=31: 21;16;12;0010016667;1610;005167;N6
2022-12-29 14:24:40.786 [update notice] sent scan-read scan.08 id QQ=31: 21;16;12;0010016667;1610;005167;N6
2022-12-29 14:24:41.551 [update notice] sent scan-read scan.08 id QQ=31: 21;16;12;0010016667;1610;005167;N6
2022-12-29 14:24:42.260 [update notice] sent scan-read scan.08 id QQ=31: 21;16;12;0010016667;1610;005167;N6
2022-12-29 14:24:42.260 [bus notice] scan 08: ;21;16;12;0010016667;1610;005167;N6
2022-12-29 14:24:45.169 [update notice] sent unknown MS cmd: 3115b5090124 / 09003231313630353030
2022-12-29 14:24:47.990 [update notice] received update-write bai SetMode QQ=10: auto;90.0;59.0;-;0;0;0;0;0;0
2022-12-29 14:24:48.662 [update notice] sent scan-read scan.15 id QQ=31: 
2022-12-29 14:24:49.384 [update notice] sent scan-read scan.15 id QQ=31: 
2022-12-29 14:24:50.106 [update notice] sent scan-read scan.15 id QQ=31: 21;16;05;0020118074;0082;005638;N7
2022-12-29 14:24:50.107 [bus notice] scan 15: ;21;16;05;0020118074;0082;005638;N7
2022-12-29 14:24:50.317 [main notice] read scan config file vaillant/15.e7c.csv for ID "e7c00", SW0206, HW7402
2022-12-29 14:24:50.393 [main notice] found messages: 401 (3 conditional on 29 conditions, 0 poll, 10 update)
```

The most important lines are these:

```
2022-12-29 14:24:27.833 [bus notice] scan 08: ;Vaillant;BAI00;0604;1303
2022-12-29 14:24:27.833 [update notice] sent scan-read scan.08  QQ=31: Vaillant;BAI00;0604;1303
2022-12-29 14:24:27.833 [bus notice] scan 08: ;Vaillant;BAI00;0604;1303
2022-12-29 14:24:32.528 [bus notice] scan 15: ;Vaillant;E7C00;0206;7402
2022-12-29 14:24:32.529 [update notice] sent scan-read scan.15  QQ=31: Vaillant;E7C00;0206;7402
2022-12-29 14:24:32.529 [bus notice] scan 15: ;Vaillant;E7C00;0206;7402
2022-12-29 14:24:34.978 [main notice] read scan config file vaillant/08.bai.csv for ID "bai00", SW0604, HW1303
2022-12-29 14:24:50.317 [main notice] read scan config file vaillant/15.e7c.csv for ID "e7c00", SW0206, HW7402
```
which indicate that a boiler 'bai' was found at address 8 and a thermostat on address 15. The configuration files `08.bai.csv` and 
`15.e7c.csv` were loaded that allow human interpretation of the messages that are sent between these two devices. Note that `ebusctl info` returns some more information about the master and slave addresses of both devices.

The configuration files that are mentioned above are hosted at https://github.com/john30/ebusd-configuration and are automatically downloaded by ebusd by default. You can also download them manually, modify them if needed and point ebusd to the path of the configuration files. See the documentation for more information.

> Edit 22/4/23:
> I used `cron` with an `@reboot` flag to start ebusd, since I like cron. However, by the time that cron fires up ebusd, the machine may not yet have an internet connection. Ebusd then fails since it cannot fetch the configuration files. This could be a good reason for storing the config files locally.

## Listening on the bus
Before controlling my boiler, I wanted to see what communication is passing on the bus between the existing Exacontrol C7 thermostat and the Thermomaster. 

You can do this by using the command `ebusctl listen`, which prints commands to the terminal at the time that they appear on the bus, e.g.:
```
bai Status01 = 70.0;54.0;-;-;-;on
bai SetMode = auto;69.5;59.0;-;0;0;0;0;0;0
bai Status01 = 70.5;54.0;-;-;-;on
bai Status01 = 70.0;54.0;-;-;-;on
```
but I found that inspecting the log `/var/log/ebusd.log` is more handy since it provides more information (timestamps and an indication of what is a read and what is a write command):

```
2022-12-29 14:35:13.034 [update notice] received unknown MS cmd: 1008b5070108 / 026400
2022-12-29 14:35:15.826 [update notice] received update-write bai SetMode QQ=10: auto;68.0;59.0;-;0;0;0;0;0;0
2022-12-29 14:35:20.833 [update notice] received read bai Status01 QQ=10: 68.0;54.0;-;-;-;on
2022-12-29 14:35:22.852 [update notice] received read bai DateTime QQ=10: nosignal;-:-:-;-.-.-;-
2022-12-29 14:35:23.107 [update notice] received read bai Status02 QQ=10: on;20;70.0;70;55.0
2022-12-29 14:35:25.862 [update notice] received update-write bai SetMode QQ=10: auto;68.0;59.0;-;0;0;0;0;0;0
2022-12-29 14:35:30.825 [update notice] received read bai Status01 QQ=10: 68.0;54.0;-;-;-;on
2022-12-29 14:35:32.802 [update notice] received update-write bai StatusCirPump QQ=10: off
2022-12-29 14:35:35.853 [update notice] received update-write bai SetMode QQ=10: auto;68.0;59.0;-;0;0;0;0;0;0
2022-12-29 14:35:40.849 [update notice] received read bai Status01 QQ=10: 68.0;54.0;-;-;-;on
```

From this output it seems that the thermostat uses the command `Status01` and `Status02` to request information from the boiler (QQ=10 is the master address of the thermostat, which is the sender of the command, [documentation here](https://github.com/john30/ebusd/wiki/4.1.-Message-definition#message-definition)).

Furthermore, the thermostat issues a write command `SetMode`, this is the interesting part! More about that later.

## Issuing a read command
So the thermostat reads data from the boiler. That sounds pretty benign, let's try to mimic that using the `read` command before attempting to write to the bus and make something go horribly wrong:

```
ebusctl read Status01
```

If you didn't wait long enough after starting ebusd, or something else went wrong, you can get the following result:
```
ebusctl read Status01
ERR: element not found
```
this means that 'Status01' is not recognised using the loaded configuration files (which may be none at all).

If everything is working properly you get something like:
```
ebusctl read Status01
41.0;40.5;-;-;-;off
```
which lists two temperatures and something that is 'off': supply and return water temperature and the circulation pump.

However, issuing this command multiple times in a row results in the same result, even when the display of the heater shows that the temperatures are changing. That's because by default the cached result is returned. Use the `-f` flag to force read from the bus:

```
ebusctl read -f Status01
```

## Issuing a write command
Writing is where things get difficult. Ebusd was designed to protect you from doing stupid things to your boiler. So proceed with care here, these instructions worked for me, but there is no guarantee that it will result in the expected outcome for you too.

Looking again at the log, we see things like:

```
2022-12-29 14:35:30.825 [update notice] received read bai Status01 QQ=10: 68.0;54.0;-;-;-;on
```

Based on this line
```
uw,,SetMode,Operation Mode,,,,00,,,hcmode,,,,flowtempdesired,,temp1,,,,hwctempdesired,,temp1,,,,hwcflowtempdesired,,temp0,,,,,,IGN:1,,,,disablehc,,BI0,,,,disablehwctapping,,BI1,,,,disablehwcload,,BI2,,,,,,IGN:1,,,,remoteControlHcPump,,BI0,,,,releaseBackup,,BI1,,,,releaseCooling,,BI2,,,,
```
in [the configuration file of hcmode](https://github.com/john30/ebusd-configuration/blob/master/ebusd-2.1.x/en/vaillant/hcmode.inc) the '68' is the desired flow temperature. So it would be tempting to try

```bash
ebusctl write -c bai SetMode '50.0;54.0;-;-;-;on'
```
based on [this post](https://github.com/john30/ebusd/issues/179#issuecomment-434628658) to try to write a set point of 50 degrees. The result is:

```bash
ERR: element not found
```
Based on [this post](https://github.com/john30/ebusd/issues/179#issuecomment-434602001) that's because the field is not intended to be written: it has the [specification](https://github.com/john30/ebusd-configuration/blob/master/ebusd-2.1.x/en/vaillant/hcmode.inc) `uw`. 

The underlying problem is that the thermostat writes `SetMode` every 10 seconds so anything that we write would be overwritten pretty soon by the thermostat. So ebusd does not allow writing to this command. However, I want my Raspberry pi to _replace_ the thermostat, so I would not have that problem. @andig and @Mateo89 suggest to create a new definition that has `wi` instead of `uw` to cope with this [here](https://github.com/john30/ebusd-configuration/issues/170#issuecomment-709592197). 

However, it's not really clear to me from these posts why one thing works and another thing does not. It's a bit black magic and that doesn't feel reassuring to go forward. At the end of the day, the messages that are sent to the boiler are really not that complex. How hard can it be to correctly mimic them? So I tried to directly write HEX messages to the boiler instead of creating a custom message definition.

## Issuing a hex command
As mentioned before, ebusd interprets command using configuration files. The data that is sent over ebus are just bits and bytes with a specific format. Let's try to mimic those instead. Some examples can be shown using `ebusctl grab`:

```
ebusctl grab result all
3108070400 / 0ab5424149303006041303 = 1: scan.08 
3108b5090127 / 094e363c3c3c3c3c3c3c = 8: scan.08 id
0315070400 / 0ab5453743303002067402 = 2: scan.15 
3115b5090127 / 094e3700000000000000 = 7: scan.15 id
1008b5100900000076ffff010100 / 0101 = 84: bai SetMode
1008b512020000 / 00 = 15: bai StatusCirPump
10feb516080054571429120422 = 15: broadcast vdatetime
1008b5040100 / 0a00ffffffffffffff0080 = 1: bai DateTime
1008b5070108 / 026400 = 15
1008b5110101 / 0952520080ffff0000ff = 2: bai Status01
1008b5110102 / 0601148c466e50 = 1: bai Status02
3115b5090124 / 09003231313630353030 = 1: scan.15 id
1008b5040100 / 0a00ffffffffffffff0080 = 30: bai DateTime
1008b5110101 / 094e4d0080ffff0000ff = 88: bai Status01
1008b5110102 / 0601148c466e50 = 30: bai Status02
1008b512020000 / 00 = 1: bai StatusCirPump
1008b51302050a / 0101 = 2
1008b5100900000076ffff010100 / 0101 = 1: bai SetMode
```
It looks like the format here is 
```<HEX command> / <HEX response> = <number of these responses> : <interpreted device> <interpreted command>```.

So the command for our previous read request `Status01` is `1008b5110101`, where `10` is the sender address, `08` is the receiver address.

Se we can do this:
```
ebusctl r -h 08b5110101
094e4c0080ffff0000ff
```
to execute the Status01 command and we receive the result `094e4c0080ffff0000ff`. Note that we had to drop the sender address for this to work since ebusd automatically adds the sender address. And we do not need to `-f(orce)` anything since hex command results are not cached.

So on to writing.. Let's decode these messages from `ebusctl grab`:
```
1008b5100900007376ffff000100/0101 = 1: bai SetMode 
1008b5100900007776ffff000100/0101 = 1: bai SetMode
1008b5100900000076ffff010100/0101 = 1: bai SetMode
```
Based on an educated guess they correspond to the following messages from the log:
```
update-write bai SetMode QQ=10: auto;57.5;59.0;-;0;0;0;0;0;0
update-write bai SetMode QQ=10: auto;59.5;59.0;-;0;0;0;0;0;0
update-write bai SetMode QQ=10: auto;0.0;59.0;-;1;0;0;0;0;0
```

So it seems that the number `73` corresponds to 57.5 degrees, since it is pretty much the only thing that changed between these hex commands. After some digging in [here](https://github.com/john30/ebusd-configuration/blob/master/ebusd-2.1.x/en/vaillant/hcmode.inc) (temp1), a link between `temp1` and `D1C` [here](https://github.com/john30/ebusd/blob/5ec5c3daa44502b3bb5a6162797cc47dfe4c5f23/src/lib/ebus/test/test_message.cpp),  and the following code
```
  // unsigned number (fraction 1/2), 0 - 100 (0x00 - 0xc8, replacement 0xff)
  add(new NumberDataType("D1C", 8, 0, 0xff, 0x00, 0xc8, 2));
```
from https://github.com/john30/ebusd/blob/bc3e6110c7f6e12d9912e09d061c9dd5f335de39/src/lib/ebus/datatype.cpp it seems that the two hex numbers are converted in a temperature using the following math:

```
temperature = (<first hex>*16 + <second hex>)/2
```
indeed, (7*16+3)/2=57.5C.

So we can substitute the temperature in the message `08b510090000<HEX TEMP HERE>76ffff000100` to write a temperature set point and to enable the pump. The response of the boiler should be `0101`:

```
ebusctl hex 08b5100900007776ffff000100
0101
```

and we'll just write `08b5100900000076ffff010100` to disable the boiler.

Note that the thermostat should be disconnected if you don't want it to overwrite your commands. 

This seems to work fine for me but without extensive testing so far. It seems that the boiler remembers this command for 15 minutes and after that timeout period it shuts itself down. So the system seems robust against a sudden disconnection of the raspberry pi.

Edit: When testing a Vaillant boiler, it did NOT shut itself down after not sending new commands to the boiler.
Edit2: It seems that the Bulex boiler also does not shut itself down. It maintains the latest temperature set point.

It was quite a challenge to piece together everything to get this working. I hope this helps someone :)

**Comment by G8GYW** [comment]
Fantastic  piece of work, thanks for making it available. It answered a number of questions that I couldn't find answers to elsewhere. I built several of those ebuzzz adapters for myself and others. It works really well on Vaillant boilers.

Mike

**Reply by ghost** [reply]
thank you @Mathadon . This Wiki contribution should be referenced on the homepage. Great stuff.

**Comment by swolenmaximus** [comment]
Terribly late with this reply but Ive been through the struggles! After an entire weekend of messing about, trying 3 different cables to connect the Shield C6 to the bus outlet of my Bulex Thermomaster and numerous firmware flashes I was ready to give up. Upon gazing at the manual (duh... RTFM!) I found a small topic that mentioned disabling the eco mode when you connect a thermostat via bus. After doing so, I managed to pull 30 something topics and pull temperature data out of the boiler! Putting this down here in case this might help a lost soul.

**Comment by hbarnard** [comment]
Thanks very much for this. I never got the earlier clunkier adapter to work (probably because of attenuation via cable length and adjustment difficulties etc. etc.) and have now managed to flash and connect the C6 to my Vaillant Ecotec. I'm planning on reading the switch-ons, maybe graphing and see what influence the thermostat settings have. I'll do a little scripting on this, if it's 'nice' I'll publish.

Not planning on doing any writing currently because it still feels like a bit of a black box. There's spec documents on scribd but out of date.
